go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -126,8 +126,14 @@ func (npw *nodePortWatcher) AddService(service *kapi.Service) {
 	addSharedGatewayIptRules(service, npw.nodeIP)
 }
 
-func (npw *nodePortWatcher) UpdateService(new, old *kapi.Service) {
-	if reflect.DeepEqual(new.Spec, old.Spec) {
+func (npw *nodePortWatcher) UpdateService(old, new *kapi.Service) {
+	if reflect.DeepEqual(new.Spec.Ports, old.Spec.Ports) &&
+		reflect.DeepEqual(new.Spec.ExternalIPs, old.Spec.ExternalIPs) &&
+		reflect.DeepEqual(new.Spec.ClusterIP, old.Spec.ClusterIP) &&
```

Comment:
It looks like we don't need to check on the ClusterIP from the types.go
```
        // clusterIP is the IP address of the service and is usually assigned                                                                                                                                                                                                   
        // randomly by the master. If an address is specified manually and is not in                                                                                                                                                                                            
        // use by others, it will be allocated to the service; otherwise, creation                                                                                                                                                                                              
        // of the service will fail. This field can not be changed through updates.                                                                                                                                                                                             
        // Valid values are "None", empty string (""), or a valid IP address. "None"                                                                                                                                                                                            
        // can be specified for headless services when proxying is not required.                                                                                                                                                                                                
        // Only applies to types ClusterIP, NodePort, and LoadBalancer. Ignored if                                                                                                                                                                                              
        // type is ExternalName.                                                                                                                                                                                                                                                
        // More info: https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies                                                                                                                                                          
        // +optional                                                                                                                                                                                                                                                            
        ClusterIP string `json:"clusterIP,omitempty" protobuf:"bytes,3,opt,name=clusterIP"`      
```
the clusterIP cannot be changed by an update. Or is this just making sure that the service is deleted and recreated?

Reply:
ah ok then we dont need to check for this here. We also use this check in other places. I'll clean it up later if that's ok.

Reply:
IIRC there are some corner cases where it can be changed, but only from "" -> valid, or from valid -> ""

Reply:
yeah, the server can mutate and clean it

Reply:
i.e. ServiceType: ClusterIP/NodePort or LoadBalancer can be converted to ServiceType: ExternalName, and the apiserver will delete the ClusterIP assigned